pallas-primitives/src/plutus_data.rs

```diff
@@ -0,0 +1,353 @@
+                // haskell implementation. See above reference.
+                encode_list(&self.fields, e, ctx)?;
+                Ok(())
+            }
+        }
+    }
+}
+
+/// Defined to encode PlutusData bytestring as it is done in the canonical
+/// plutus implementation
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[serde(into = "String")]
+#[serde(try_from = "String")]
+pub struct BoundedBytes(Vec<u8>);
+
+impl From<Vec<u8>> for BoundedBytes {
+    fn from(xs: Vec<u8>) -> Self {
+        BoundedBytes(xs)
+    }
+}
+
+impl From<BoundedBytes> for Vec<u8> {
+    fn from(b: BoundedBytes) -> Self {
+        b.0
+    }
+}
+
+impl Deref for BoundedBytes {
+    type Target = Vec<u8>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl TryFrom<String> for BoundedBytes {
+    type Error = hex::FromHexError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        let v = hex::decode(value)?;
+        Ok(BoundedBytes(v))
+    }
+}
+
+impl From<BoundedBytes> for String {
+    fn from(b: BoundedBytes) -> Self {
+        hex::encode(b.deref())
+    }
+}
+
+impl fmt::Display for BoundedBytes {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let bytes: Vec<u8> = self.clone().into();
+
+        f.write_str(&hex::encode(bytes))
+    }
+}
+
+impl<C> Encode<C> for BoundedBytes {
+    fn encode<W: minicbor::encode::Write>(
+        &self,
+        e: &mut minicbor::Encoder<W>,
+        _: &mut C,
+    ) -> Result<(), minicbor::encode::Error<W::Error>> {
+        // we match the haskell implementation by encoding bytestrings longer than 64
+        // bytes as indefinite lists of bytes
+        const CHUNK_SIZE: usize = 64;
+        let bs: &Vec<u8> = self.deref();
+        if bs.len() <= 64 {
+            e.bytes(bs)?;
+        } else {
+            e.begin_bytes()?;
+            for b in bs.chunks(CHUNK_SIZE) {
+                e.bytes(b)?;
+            }
+            e.end()?;
+        }
+        Ok(())
+    }
+}
+
+impl<'b, C> minicbor::decode::Decode<'b, C> for BoundedBytes {
+    fn decode(d: &mut minicbor::Decoder<'b>, _: &mut C) -> Result<Self, minicbor::decode::Error> {
+        let mut res = Vec::new();
+        for chunk in d.bytes_iter()? {
+            let bs = chunk?;
+            res.extend_from_slice(bs);
+        }
+        Ok(BoundedBytes::from(res))
+    }
+}
+
+fn encode_list<C, W: minicbor::encode::Write, A: minicbor::encode::Encode<C>>(
+    a: &Vec<A>,
+    e: &mut minicbor::Encoder<W>,
+    ctx: &mut C,
+) -> Result<(), minicbor::encode::Error<W::Error>> {
+    // Mimics default haskell list encoding from cborg:
+    // We use indef array for non-empty arrays but definite 0-length array for empty
+    if a.is_empty() {
+        e.array(0)?;
+    } else {
+        e.begin_array()?;
+        for v in a {
+            e.encode_with(v, ctx)?;
+        }
+        e.end()?;
+    }
+    Ok(())
+}
```

Comment:
I don't mind this, but maybe better suited for pallas::codec::utils as a newtype (`HardanoList` or better name)?

Reply:
I have intention to rewrite a bit some of the data encoder here; because the underlying type for `Constr` in particular has proven quite unpractical to use. I understand why the implementation ended that way, but I believe we can do better! 

So for now, I've only parked the Plutus Data-related code in a separate module but haven't tweaked anything. So, this code has just been moved around. This encoding in particular is specific to Plutus data and seen nowhere else in the Cardano codebase; so I believe it better lives with the rest of the Plutus Data stuff.

Reply:
I thought this was a new fn.
Ok, lets keep it until a further refactor.